torch/csrc/jit/topological_index.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <c10/util/Exception.h>
+#include <map>
+#include <unordered_map>
+
+namespace torch {
+namespace jit {
+
+template <typename T>
+struct TopologicalIndex;
+
+struct Node;
+using node_topological_index = TopologicalIndex<Node*>;
+using topo_position_t = uint64_t;
+
+// Index to track a topological ordering of nodes. Owned by a block,
+// representing the nodes in the block.
+//
+// This lets us answer questions like "is this node before another node"
+// efficiently, which is useful for optimization. It should be kept up to date
+// with node insertions/deletions by the owning block.
+//
+// The basic scheme is: nodes are assigned topological indices within a
+// 64-bit space. Appending a node moves assigns a position that's a big interval
+// higher than the last node, giving room for insertions in between. If we ever
+// run out of room, we rebuild the index.
+//
+// NOTE: this relies on some implementation details of the node listing, so
+// don't reuse without checking those assumptions.
+template <typename T>
+struct TopologicalIndex {
+ private:
+  // The node list is implemented as a circular linked list, with the output
+  // node as the head/tail. Therefore, the index needs to treat "appends" to the
+  // output as appends to the input.
+  T input_;
+  T output_;
+
+  // Lower and upper bounds of the index. Inclusive range.
+  const topo_position_t lowerBound_;
+  const topo_position_t upperBound_;
+
+  // How far away to space nodes that are appended to the graph.
+  // should be 2^n, where:
+  //   - n is the maximum number of repeated insertions without a re-index
+  //   - 2^(64-n) is the maximum number of appends to the end without reindex
+  const topo_position_t defaultInterval_;
+
+  std::map<topo_position_t, T> positionToObj_;
+
+ public:
+  TopologicalIndex(T input, T output)
+      : TopologicalIndex(
+            input,
+            output,
+            0,
+            UINT64_MAX,
+            1099511627776ULL /* 2^40 */) {}
+
+  // This constructor is for tests only, so we can test boundary conditions.
+  TopologicalIndex(
+      T input,
+      T output,
+      topo_position_t lowerBound,
+      topo_position_t upperBound,
+      topo_position_t defaultInterval)
+      : input_(input),
+        output_(output),
+        lowerBound_(lowerBound),
+        upperBound_(upperBound),
+        defaultInterval_(defaultInterval) {
+    AT_ASSERT(upperBound_ > lowerBound_);
+
+    setPos(input_, lowerBound_);
+
+    // Don't put the output node in the index, since that would prevent us from
+    // appending efficiently. Instead just give it the max index for comparison
+    output_->topo_position_ = upperBound_;
+  }
+
+  TopologicalIndex(const TopologicalIndex&) = delete;
+  void operator=(const TopologicalIndex&) = delete;
+
+  // Insert `toInsert` after `insertPoint` in the topological index
+  void insertAfter(T insertPoint, T toInsert) {
```

Comment:
There is no need for any auxillary data structures like positionToObj_. This function can be implemented as:

```
void assignTopoPosition(Node* just_inserted) {
    auto return_node = just_inserted->owningBlock()->return_node();
    // append
    if(just_inserted->next() == return_node) {
      // appending to the end of the list:
      toInsert->topo_position = upperBound_ += interval;
      if(upperBound_ wrapped around) {
        renumber();
      }
    // prepend
    } else if(just_inserted->prev() == return_node) {
      // prepend list:
      toInsert->topo_position = lowerBound_ -= interval;
      if(lowerBound_ wrapped around) {
        renumber();
      }
    } else {
      Node* prev = insert_point->prev();
      Node* next = insert_point->next(); 
      if(prev->topo_position + 1 == next->topo_position)
        renumber(); 
      just_inserted->topo_position = (prev->topo_position + next->topo_position) / 2;
    }
  }
```